Onfleet node

0xsemaj opened the pull request:
### Preparation

- [x]  Set up your editor for code formatting (indentation, new lines, linting). If you use Visual Studio Code, you can use the TSLint extension for linting.
- [x]  Get credentials (e.g., Client ID, Client Secret, API key, user login, user password, website URL) for the service you are building a node for.

### Development

- [x] Open a pull request as early as possible with WIP in the pull request title. (We're doing this in one big PR 😆 )
- [x] If you are creating a node requested by a community member, make sure to comment on the feature request in the community forum (opens new window).
- [x] Ensure complementary operations to each resource (e.g., `create`, `delete`) have been added.
- [x] Ensure the node works with multiple items via one input.
- [x] Ensure the parameters have the correct type.
- [x] Mind the defaults: if the service has a default as `true`, keep it as `true`. Changing default values can break the existing workflows of the users.
- [x] Check if the node disposes of everything properly, in particular, if connections were properly closed.
- [x] Check your code using Nodelinter to ensure a clean lint before submitting your pull request

### Testing

- [x] Test `create` and `update` operations with all fields/operations.
- [x] Test the `continueOnFail` option with a Function node. (For example, a Widget node has a GET operation that takes a widgetId and returns information on the widget. To test that the workflow continues on fail, set the Widget node to continue on fail, create a Function node, return a valid and an invalid widgetId, connect the Function node to Widget node, and run the workflow. The Widget node should show two items: one with information on the widget and another one with the error from having passed an invalid ID.)

### Code formatting
- [x] Ensure the branch lints cleanly by running npm run lint.
- [x] Ensure the indentation is correct. Check this in the `editorconfig`.
- [x] Ensure there are no extra spaces. Check this in the `editorconfig`.
- [x] Code comment dividers inside if-branches.
- [x] Use "create/delete" verbs for operations, except for tags, where you should use "add/remove". (We still opt to use `create` in several cases to be uniform with our API documentation)

### Errors and Outputs
- [x] Ensure empty API responses return { success: true }.
- [x] Ensure the error responses are handled and displayed correctly (e.g., malformed requests, requests with invalid credentials) and use the current format. You can check this by making failing requests to the API.
- [x] Check if the response can be simplified and add a simplify function (e.g., SecurityScorecard node). (Not needed in our case)
- [x] Ensure the response from `Create` is consistent with `Get`.
- [x] Ensure the response from `Get All` is consistent with `Get`.

### Presentation
- [x] Ensure the primary menu contains only required parameters.
- [x] Ensure a JSON object is not shown in a single column in Table view. (Nested JSON)
- [x] Make sure all `GetAll` operations have the fields return and limit.
- [x] Ensure the property subtitle is set.
- [x] Make sure the pagination (if any) is working correctly. Set Limit 1.

### Writing
- [x] Ensure all descriptions are correct and end with a period.
- [x] Ensure that most descriptions exist, excluding redundant ones.
- [x] Ensure IDs in `displayNames` are capitalized (i.e.: "IDs", not "ids" or "Ids").
- [x] Ensure that IDs, if multiple, have descriptive qualifiers.
- [x] Ensure the name property in `description` in the node class is written in camelCase.
- [x] Ensure the file name and the Class name are identical.

### Branding
- [x] Ensure the name of the service is written correctly (e.g., "GitHub" not "Github"). If the node is a trigger node, ensure it is named as such, by adding "Trigger" after the service name (e.g., "Trello Trigger").
- [x] Ensure the logo is either a PNG or SVG, ideally the latter. Vecta is a good website to find SVGs of different applications.
- [x] If the logo is an SVG, ensure the canvas is a perfect square. If the logo is PNG, ensure it is 60x60 pixels and compressed.
- [x] Ensure the border color of the node matches the branding of the service.

### Nice-to-haves (optional)
- [x] Add handler for `continueOnFail`. This feature is included in some of the newest nodes (e.g Lemlist node) to continue the workflow even if the node's execution fails.
- [x] Remove required: `false` and `description: ''` in the node descriptions (e.g., Lemlist node).
- [x] At call site, specify first `body` and then `qs`.
- [x] At call site, prepend the endpoint with slash `/` (e.g., "/campaign").

Commit:
feat: added Onfleet nodes

Added Onfleet nodes for working with different endpoints like:
organizations, administrators, workers, hubs, teams, destinations, recipients,
containers and webhooks.

Commit:
style: fixed typos, arrays uniformity, unnecesary files

Commit:
refactor: changed add to create in comments and labels

0xsemaj committed:
Merge pull request #1 from YokySantiago/master

INT-441: Add Onfleet nodes to n8n

Commit:
feat: added name field to onfleet trigger node

0xsemaj committed:
Merge pull request #2 from YokySantiago/master

feat: added name field to onfleet trigger node

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=2593) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=2593) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: yoky-devsavant<br/>:x: jamesliupenn<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=2593) it.</sub>

jamesliupenn commented:
@ivov hi there, do you know what's the typical lead time to get a new node added? We're anticipating this release and hope to get an estimate of when this will be live.

janober commented:
Thanks a lot for the contribution @jamesliupenn. The node is one of the next ones that will get reviewed. So should not be much longer.

jamesliupenn commented:
Thanks! I realized we missed a few operations so we'll be making some updates tomorrow

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts:
```diff
@@ -0,0 +1,172 @@
+} as INodeProperties;
+
+const hubField = {
+	displayName: 'Hub',
```
Hub ID. Can you load all the available Hubs?

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts:
```diff
@@ -0,0 +1,172 @@
+} as INodeProperties;
+
+export const teamFields = [
+	{
```
Team ID

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts:
```diff
@@ -0,0 +1,172 @@
+		required: true,
+	},
+	{
+		displayName: 'Additional fields',
```
Additional Fields

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/TeamDescription.ts:
```diff
@@ -0,0 +1,172 @@
+		],
+	},
+	{
+		displayName: 'Additional fields',
```
Additional Fields

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WebhookDescription.ts:
```diff
@@ -0,0 +1,131 @@
+} from 'n8n-workflow';
+import { webhookMapping } from '../WebhookMapping';
+
+export const webhookOperations = [
```
Why is this resource needed when the trigger is already handling these endpoints?

RicardoE105 replied:
I would remove it complely

YokySantiago replied:
Well, the trigger node is a listener, however, we'd like to run webhook operations for other apps from n8n nodes.  i.e: 

- We want to be able to create an Onfleet webhook but n8n is not the listener
- We want to delete an Onfleet webhook however n8n is not the listener
- We want to list created Onfleet webhooks and they aren't necessary of n8n

In addition, this is part of the API offering and we'd like to keep them for uniformity.

jamesliupenn replied:
There are some use cases where n8n is used as middleware and not necessarily the webhook server. Hence we would like to keep the current setup so the webhook resource is handling endpoints outside of the trigger node.

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+				value: 'create',
+				description: 'Create a new Onfleet worker.',
+			},
+			{
```
Delete

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+				description: 'Remove an Onfleet worker.',
+			},
+			{
+				name: 'List',
```
Get All

RicardoE105 replied:
Pagination is missing

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+				description: 'List all Onfleet workers.',
+			},
+			{
+				name: 'List workers by location',
```
Can we merge this operation and the getAll operation?

RicardoE105 replied:
As long as they report the same data I think they can be merged.

RicardoE105 replied:
Pagination is missing

RicardoE105 replied:
The collection needs to be called filters.

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+				name: 'Get',
+				value: 'get',
+				description: 'Get a specific Onfleet worker.',
+			},
```
Sort alphabetically

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+		},
+		{
+			name: 'Motorcycle',
+			value: 'MOTORCYCLE',
```
Sort alphabetically

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+} as INodeProperties;
+
+const teamsField = {
+	displayName: 'Teams (JSON)',
```
No need to do a JSON. Use string with mutlipleValues set to true

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+		{
+			name: 'Off-duty',
+			value: 0,
+		},
```
Sort alphabetically

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+const filterField = {
+	displayName: 'List of fields to return',
+	name: 'filter',
+	type: 'string',
```
Can we load all the possible fields so they do not have to type them in?

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+	displayName: 'Date',
+	name: 'date',
+	type: 'dateTime',
+	default: Date.now(),
```
Set default to empty

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+
+const scheduleTimezoneField = {
+	displayName: 'Timezone',
+	name: 'timezone',
```
Load all timezones from moment.

			async getTimezones(
				this: ILoadOptionsFunctions,
			): Promise<INodePropertyOptions[]> {
				const returnData: INodePropertyOptions[] = [];
				for (const timezone of moment.tz.names()) {
					const timezoneName = timezone;
					const timezoneId = timezone;
					returnData.push({
						name: timezoneName,
						value: timezoneId,
					});
				}
				return returnData;
			},

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+	displayName: 'Start',
+	name: 'start',
+	type: 'dateTime',
+	default: Date.now(),
```
Set default to ''

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+	displayName: 'End',
+	name: 'end',
+	type: 'dateTime',
+	default: Date.now(),
```
Set default to ''

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+
+export const workerFields = [
+	{
+		displayName: 'ID',
```
Worker ID

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+			vehicleColorField,
+		],
+	},
+	{
```
Title case

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+		],
+	},
+	{
+		displayName: 'Additional fields',
```
Update Fields

RicardoE105 commented on packages/nodes-base/nodes/Onfleet/descriptions/WorkerDescription.ts:
```diff
@@ -0,0 +1,529 @@
+	{
+		displayName: 'Additional fields',
+		name: 'additionalFields',
+		type: 'collection',
```
The analytics fields need to be optional.

RicardoE105 replied:
Rename the collection to Options.

RicardoE105 reviewed (commented):
Got reviewed. Let me know if you have questions about any comments. @jamesliupenn

Commit:
feat: added team endpoints to onfleet node

Added team auto-dispatch and driver time estimate endpoints to Onfleet
node

Commit:
style: remove dots in descriptions and fixed some typos

0xsemaj committed:
Merge pull request #3 from YokySantiago/master

INT-483: Add missing endpoints to n8n-onfleet node

jamesliupenn commented:
We'll be prioritizing this in the ongoing sprint, thanks for the feedback @RicardoE105, we'll keep you posted once we finish developing it internally.

YokySantiago commented on packages/nodes-base/nodes/Onfleet/descriptions/ContainerDescription.ts:
```diff
@@ -0,0 +1,188 @@
+		},
+		options: [
+			{
+				name: 'Insert tasks (or append)',
```
Hey @RicardoE105 !
In this case we don't have container information. This is the meaning of container for us "Containers are an abstraction which describes task assignment. A container is an ordered list of tasks. Organizations, teams and workers all correspond to containers as they can all be assigned tasks." So, when we get information about a container, we need to select the type (Organization, Team, Worker) in order to get the task assignment for the container selected.
So, We would like to keep it all together as a resource

YokySantiago commented on packages/nodes-base/nodes/Onfleet/OnfleetTrigger.node.ts:
```diff
@@ -0,0 +1,224 @@
+				const webhookData = this.getWorkflowStaticData('node') as IDataObject;
+				const credentials = await this.getCredentials('onfleetApi') as ICredentialDataDecryptedObject;
+				const event = this.getNodeParameter('event', 0) as string;
+				const { name = '' } = this.getNodeParameter('additionalFields', 0) as IDataObject;
```
Hey Ricardo! Well, this field is part of the API offering and we'd like to keep them for uniformity

jamesliupenn replied:
I would have to agree on keeping the name, many of our customers use the name as an indicator of what types of webhooks it is and where it is utilized. Using n8n-url would simply be too generic

jamesliupenn commented on packages/nodes-base/nodes/Onfleet/descriptions/OrganizationDescription.ts:
```diff
@@ -0,0 +1,47 @@
+			},
+			{
+				name: 'Get delegatee details',
+				value: 'getDelegatee',
```
We would like things to still relate to our API docs for better support, especially on the language used. In this case, delegatee has always been used and the change in terminology would simply confuse our users.

jamesliupenn commented on packages/nodes-base/nodes/Onfleet/descriptions/TaskDescription.ts:
```diff
@@ -0,0 +1,380 @@
+				description: 'Create a new Onfleet task.',
+			},
+			{
+				name: 'Create multiple tasks',
```
Batch create and create are two different operations in our API, they should both be supported in cases where some users need to bulk upload tasks versus when someone stream processes task creation

jamesliupenn commented on packages/nodes-base/nodes/Onfleet/descriptions/TaskDescription.ts:
```diff
@@ -0,0 +1,380 @@
+				description: 'Remove an Onfleet task.',
+			},
+			{
+				name: 'List',
```
_from_ is not optional when listing all tasks, it is actually required https://docs.onfleet.com/reference#list-tasks

jamesliupenn commented on packages/nodes-base/nodes/Onfleet/descriptions/TaskDescription.ts:
```diff
@@ -0,0 +1,380 @@
+				description: 'The state of the tasks.',
+			},
+			{
+				displayName: 'LastId',
```
Yes this is needed for pagination in `/tasks/all` endpoint

Commit:
feat: added fixes according to comments made on the n8n PR

added new fixed collections, refactored the code according to comments
made on the n8n pr

Commit:
fix: fixed recipient and destination cretion

Commit:
docs: added docstrings for format some functions

added docstrings for new functions addded for formatting the destination
and recipient objects

Commit:
style: formatting the code according to n8n nodelinter

0xsemaj committed:
Merge branch 'n8n-io:master' into master

Commit:
fix: typos and better descriptions

0xsemaj committed:
Merge pull request #4 from YokySantiago/master

Refactor according to n8n PR comments

jamesliupenn commented:
@RicardoE105 this is ready for another round of review as we have resolved the issues you mentioned. Please keep us posted👍👍

RicardoE105 commented:
@jamesliupenn perfect. Will review it again as soon as I can.

RicardoE105 commented:
@jamesliupenn had a chance to review again. This time I did a deeper review than last time and found the following. Can you please address these issues? Thanks.

Onfleet review

Admin resource

Admin:getAll

- [ ] Admin:getall missing pagination

Admin:delete

- [ ] Return { success: true } when the operation is successfull

Task resource

- [ ] Do only task:create and do the batching behind the scenes. Batching if available should be the implementation to use.

Task:create 

- [ ] Every field that is not required should be under a collection. I was able to create a task only with the destination address, which means, that recipients it’s not required
- [ ] When setting the address you have a collection called “additional fields” that is not necessary. Add them at the same level of the destination address and make them not required. Do the same with `Recipient`
- [ ] Rename additional task fields to Additional Fields

Task:clone
- [ ] If the operation does not return anything when it ran successfully return { success: true }
- [ ] Rename collection options to additional fields
- [ ] Move what is inside the collection override to the same level the options parameters are

Task:complete
- [ ] If this operation completes the task why would I make the success parameter false?
Return { success: true } when the operation is successful

Task:delete
- [ ] Return { success: true } when the operation is successfull

Task:getAll
- [ ] Missing pagination
- [ ] From does not seem to be a required parameter. Add It under the collection filter fields
- [ ] Rename filter fields collection to filters
- [ ] Remove Last ID parameter.
- [ ] It does not seem to return the task I created

Task:get
- [ ] Is the short id parameter really needed? Can just detect that behind the scenes and make the corresponding call?


Container resource
- [ ] I would probably only leave the get operation for this resource. Then, move insert and update to another resource called containerTask.

Container:insertTasks

- [ ] I would create different operations depending on the insert type. for example: containerTask:append, containerTask:Prepend.
- [ ] Rename tasks to Task IDs

Hub resource

Hub:getAll
- [ ] Missing pagination

Hub:update
- [ ] Destination should be under update fields

Team resource

Team:auto-dispatch

- [ ] What is road end exactly? an address?
- [ ] Return { success: true } when the operation is successfull

Team:create

- [ ] All the fields marked as required, are not actually required. Everything that is not required should not be at the main level but instead should be moved underneath a collection. In this case, the additional fields collection

Team:delete
- [ ] Return { success: true } when the operation is successfull

Team:getAll

- [ ] Missing pagination

Team:getTimeStimates

- [ ] Dropoff and pickup do not seem to be required parameters. Both need to be under additional fields if this is the case

Webhook resource

- [ ] Is it the webhook resource usufull in the node? I would not say so, since we are already using those endpoints in the trigger. Do you have a use case in mind where you need to create Webhooks dynamically? 

Worker resource

Worker:create
- [ ] It does not seem to be working. No matter the input I provide.

worker:getAll
- [ ] Missing pagination

YokySantiago committed:
[INT-510] n8n: Address additional problems from n8n code review (#5)

* Fixed some error creating a worker, moving some fields under additional fields collection

* Fixed returned values for delete operations, making some changes for style code

* Added operational error since required property is not working for dateTime fields

jamesliupenn commented:
**Admin resource**
Admin:getAll

- [ ] Admin:getall missing pagination

Admin:delete

- [ ] Return { success: true } when the operation is successful

**Task resource**

- [ ]  Do only task:create and do the batching behind the scenes. Batching if available should be the implementation to use.

Task:create
- [ ] Every field that is not required should be under a collection. I was able to create a task only with the destination address, which means, that recipients it’s not required
- [ ]  When setting the address you have a collection called “additional fields” that is not necessary. Add them at the same level of the destination address and make them not required. Do the same with Recipient
- [ ] Rename additional task fields to Additional Fields

Task:clone
- [ ]  If the operation does not return anything when it ran successfully return { success: true }
- [ ]  Rename collection options to additional fields
- [ ]  Move what is inside the collection override to the same level the options parameters are

Task:complete
- [ ]  If this operation completes the task why would I make the success parameter false?
- [ ] Return { success: true } when the operation is successful

Task:delete
- [ ]   Return { success: true } when the operation is successful

Task:getAll
- [ ]  Missing pagination
- [ ]  From does not seem to be a required parameter. Add It under the collection filter fields
- [ ] Rename filter fields collection to filters
- [ ]  Remove Last ID parameter.
- [ ]  It does not seem to return the task I created

Task:get
- [ ]  Is the short id parameter really needed? Can just detect that behind the scenes and make the corresponding call?

**Container resource**
- [ ]  I would probably only leave the get operation for this resource. Then, move insert and update to another resource called containerTask.

Container:insertTasks
- [ ]  I would create different operations depending on the insert type. for example: containerTask:append, containerTask:Prepend.
- [ ]  Rename tasks to Task IDs

**Hub resource**
Hub:getAll
- [ ] Missing pagination

Hub:update
- [ ]   Destination should be under update fields

**Team resource**
Team:auto-dispatch
- [ ]  What is road end exactly? an address?
- [ ]  Return { success: true } when the operation is successful

Team:create
- [ ]  All the fields marked as required, are not actually required. Everything that is not required should not be at the main level but instead should be moved underneath a collection. In this case, the additional fields collection

Team:delete
- [ ] Return { success: true } when the operation is successful

Team:getAll
- [ ]   Missing pagination

Team:getTimeStimates
- [ ]  Drop Off and Pick Up do not seem to be required parameters. Both need to be under additional fields if this is the case

**Webhook resource**
- [ ] Is it the Webhook resource useful in the node? I would not say so, since we are already using those endpoints in the trigger. Do you have a use case in mind where you need to create Webhooks dynamically?

**Worker resource**
Worker:create
- [ ]  It does not seem to be working. No matter the input I provide.

worker:getAll
- [ ]   Missing pagination

jamesliupenn commented:
@YokySantiago to finish up the checklist items, @RicardoE105 this is over to you for review

YokySantiago commented:
In order to complete what James mentioned

 **Task:getAll**

- [ ] It does not seem to return the task I created. Are you testing with an Onfleet account? Did you set a From value? From field is required for this operation but n8n it's not displaying it as required so, we created a new topic in n8n and in the meantime we added a manual validation. [n8n report](https://community.n8n.io/t/datetime-field-marked-as-required-is-not-working/11383)

**Container resource**
In this case we don't have container information. This is the meaning of container for us "Containers are an abstraction which describes task assignment. A container is an ordered list of tasks. Organizations, teams and workers all correspond to containers as they can all be assigned tasks." So, when we get information about a container, we need to select the type (Organization, Team, Worker) in order to get the task assignment for the container selected.
So, We would like to keep it all together as a resource

**Team:auto-dispatch**

- [ ]  What is road end exactly? an address?. This field is a select now with some additional option, improving the usability. **FIXED**

**Team:create**

- [ ] All the fields marked as required, are not actually required. Everything that is not required should not be at the main level but instead should be moved underneath a collection. In this case, the additional fields collection. **FIXED**

jamesliupenn commented:
> @jamesliupenn @YokySantiago I'm trying to test the team resource, but it says "payment is required". Can you provide a test account with access to the team endpoints?
> 
> <img width="1147" alt="image" src="https://user-images.githubusercontent.com/16496553/154765080-71da2a72-5dc0-42bf-870e-0f1ef5c063a1.png">
> 
> Also, to create a container I need a container ID. Where can I get a container ID from? 

What endpoint on the teams endpoint are you trying to access? If you are hitting specific endpoints like auto dispatch, you will need that feature enabled on your account and I can help you with that.

Containers are abstractions of task assignments, so if you assign a task to a worker/org/team, there will be a container created. The container IDs are be available via the worker/org/team ID, so basically depending on which entity you are accessing the container for, use that entity ID.

An example would be accessing a team's container, simply use that team ID

jamesliupenn commented:
@RicardoE105 you should be on a test account and has the auto-dispatch endpoint enabled now

janober commented:
Thanks a lot for your contribution. Got merged with #2845

janober commented:
Got released with `n8n@0.165.0`
